### PR TITLE
spring-cloud-storage: Added GcsLastModifiedFileListFilter

### PR DESCRIPTION
**Pull request by hosainnet**
To be able to filter for newly added files, or files older than x relative to the current system time

Fixes https://github.com/spring-cloud/spring-cloud-gcp/issues/2467

Largley based on https://github.com/spring-projects/spring-integration/blob/master/spring-integration-file/src/main/java/org/springframework/integration/file/filters/LastModifiedFileListFilter.java

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@hosainnet Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2468)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2468) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@hosainnet Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2468)!

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	private Consumer<BlobInfo> discardCallback;
+
+	/**
+	 * Construct a {@link GcsLastModifiedFileListFilter} instance with provided {@link #age}.
+	 * Defaults to 60 seconds.
+	 *
+	 * @param age the age value.
+	 * @param unit the uunit for the age value.
+	 */
+	public GcsLastModifiedFileListFilter(long age, TimeUnit unit) {
```
Comment (meltsufin) [inline spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java:55]: Can you use `java.time.Duration` instead?

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	public void addDiscardCallback(@Nullable Consumer<BlobInfo> discardCallbackToSet) {
+		this.discardCallback = discardCallbackToSet;
+	}
+
+	@Override
+	public List<BlobInfo> filterFiles(BlobInfo[] files) {
+		List<BlobInfo> list = new ArrayList<>();
+		long now = System.currentTimeMillis() / ONE_SECOND;
+		for (BlobInfo file : files) {
+			if (fileIsAged(file, now)) {
```
Comment (meltsufin) [inline spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java:69]: This if-else-if is duplicated with the `accept()` method. Can you try refactoring?

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java (hunk 1 above)
Comment (dzou) [reply spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java]: Done.

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java (hunk 2 above)
Comment (dzou) [reply spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/integration/filters/GcsLastModifiedFileListFilter.java]: Done.

**Comment by dzou** [comment]
@hosainnet @meltsufin - I added some polish commits to the PR.

I also tried clarifying the wording of the javadoc -

Does this accurately describe the filter?

```
 * The {@link GcsLastModifiedFileListFilter} is a filter which excludes all files
 * that were updated within a specified amount of time.
 *
 * <p>More specifically, it excludes all files whose {@link BlobInfo#getUpdateTime()} is
 * within {@link #age} of the current time.
```

**Comment by elefeint** [comment]
@hosainnet Could you help us with the filter's intent? Is the requirement to process "files older than X" or "files newer than X"? The issue #2467 seems to describe processing new files, but Spring Integration's `LastModifiedFileListFilter` class is meant to process older files to avoid processing the unstable still-being-written ones ([JIRA link to the original issue](https://jira.spring.io/browse/INT-3764)).

Your filter is useful either way, but it affects how we document it.

**Comment by hosainnet** [comment]
> @hosainnet Could you help us with the filter's intent? Is the requirement to process "files older than X" or "files newer than X"? The issue #2467 seems to describe processing new files, but Spring Integration's `LastModifiedFileListFilter` class is meant to process older files to avoid processing the unstable still-being-written ones ([JIRA link to the original issue](https://jira.spring.io/browse/INT-3764)).
> 
> Your filter is useful either way, but it affects how we document it.

Thanks all for your input.

@elefeint I think I mixed things up a little sorry so let me explain - my use case is actually what #2467 describes, i.e ignore existing files, and only consider new ones from _now_ on (where now is when the app starts)

I was hoping that implementing `LastModifiedFileListFilter.java` would solve that scenario but having tested the current PR implementation again now it clearly doesn't, as its purpose is processing files that have been there for x duration old, which is not the same.

What I currently have in production which works (3rd filter):

```java
        final long localDateTime = LocalDateTime.now().toInstant(OffsetDateTime.now().getOffset()).toEpochMilli();
        return new ChainFileListFilter<>(
                List.of(
                        new GcsSimplePatternFileListFilter("some-pattern/file.csv"),
                        new GcsPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), "some-pattern/file.csv"),
                        new DiscardAwareFileListFilter<>() {
                            @Override
                            public List<BlobInfo> filterFiles(BlobInfo[] files) {
                                //discard old files
                                return Arrays.stream(files)
                                        .filter(blobInfo -> blobInfo.getUpdateTime() > localDateTime)
                                        .collect(Collectors.toList());
                            }

                            @Override
                            public void addDiscardCallback(Consumer<BlobInfo> discardCallback) {
                                //no-op
                            }
                        }
                )
        );
```

Perhaps I should raise a new PR with a new `GcsAcceptFromPointInTimeFileListFilter` (for lack of a better name) where `new GcsAcceptFromPointInTimeFileListFilter()` returns files that where update time >= the time of instance creation, and `new GcsAcceptFromPointInTimeFileListFilter(instant)` would do the same from but relative to provided Instant object. Do you all think that might be a useful filter to have here?

Regarding this PR's filter implementation, if that's something you'd like to keep, happy with your new javadoc @dzou or any other changes you'd like me to make

**Comment by dzou** [comment]
@hosainnet - Thanks! That's ok, we appreciate all these contributions. I'll merge this one in and then we'll also review the other filter you submitted.

**Comment by hosainnet** [comment]
> @hosainnet - Thanks! That's ok, we appreciate all these contributions. I'll merge this one in and then we'll also review the other filter you submitted.

Thank you!

**Pull request merged**